utils/build_swift/build_swift/driver_arguments.py

```diff
@@ -594,6 +594,10 @@ def create_argument_parser():
                 "for each cross-compiled toolchain's destdir, useful when building "
                 "multiple toolchains and can be disabled if only cross-compiling one.")
 
+    option('--custom-native-toolchain', store,
+           default=None,
+           help="Specify a custom location to find compilers and tools")
```

Comment:
Interesting, how is this intended to differ from the existing `--native-{swift,clang,llvm}-tools-path` flags?

Reply:
@buttaface I think that with time, we can eliminate the usage of that flag in favor of the custom toolchain option. I don't think that we ever specify them as being in different places.

@etcwilde it would be great if you could transition those as well in a subsequent PR. I think it would be pretty easy (just change the build-preset file that specifies this to specify custom native toolchain). It would be a better end product, but if you are worried about time, I am fine with adding a TODO even though I would prefer otherwise.

Reply:
@buttaface I had looked at using those, and from the docs they seemed like they should have worked.
Maybe I missed it, but I didn't see how those affected the compiler selected for python products (Toolchain::find_tools doesn't seem to consider it at all), and on log inspection, I was picking up the wrong (host, not toolchain) compiler for some of the build products, so I'd figured I didn't fully understand the semantics of the flags. I think longer-term, you may actually want a proper sterile workspace for all of the tools. 

@gottesmm Yeah, I can do that. Like you said, they could come from different places, though they probably should actually be coming from the same place to ensure that the tools are compatible.

Reply:
Those three flags to set paths to prebuilt LLVM, clang, or Swift tools have [been in `build-script-impl` since the open source release](https://github.com/apple/swift/commit/4ac9c80809450e4566946f40bbdb639b2ba745d8#diff-1a5a064af373f742294e3b312a5c8d0ab52f5aef78052490e6005165472cda48R146), to which [I added these Python build-script flags so we could build Python products like SPM too](https://github.com/apple/swift/pull/34437/files#diff-d7765215259ca629c1a1910c1f60768d4e5c9abafd87f624f97f0d3eff33960cR377).

@etcwilde, since those old bash flags were simply bridged over into Python by me, I was unaware of `Toolchain::find_tool` and since all the build products I cared about worked or required other changes anyway, hadn't looked at it. If you want to consolidate these old flags and tie them into the Python build process better, just let us know how you want to proceed, but I think adding yet another flag here on top of the three existing flags would be a mistake.

Reply:
That's what I had thought from the docs.

Anyway, the llvm flag connects up with `SWIFT_NATIVE_LLVM_TOOLS_PATH` in cmake. That variable only seems to get set, not used, so I think `--native-llvm-tools-path` doesn't actually do anything.

`SWIFT_NATIVE_CLANG_TOOLS_PATH` sets `CMAKE_CXX_COMPILER` and `CMAKE_C_COMPILER` in the stdlib and runtime build, so that does actually have an effect, though is a straightforward.

`SWIFT_NATIVE_SWIFT_TOOLS_PATH` is the only one that is non-trivial to replace from a quick glance, though I suspect it would only take a bit of cleanup to get it to the point of setting `CMAKE_Swift_COMPILER` appropriately.
A quick search has it showing up in localization and the bootstrapping mode on non-darwin platforms. Since the new flag does a walk of the directories to find the tools, the original directories passed to these flags will still work in a migration.

So anyway, I agree, there shouldn't be four flags that are all for effectively the same purpose. I would like to split the adding of the new flag and removal of the old flags into separate PRs though.

Reply:
>  so I think --native-llvm-tools-path doesn't actually do anything.

It used to be [wrongly used in place of the clang path, which I removed](https://github.com/apple/swift/pull/34437/files#diff-16566c91343a1660010c4c24db2b29a4fb1375cee942b57e04b194acce6c6861L105), so it is currently unused.

> So anyway, I agree, there shouldn't be four flags that are all for effectively the same purpose. I would like to split the adding of the new flag and removal of the old flags into separate PRs though.

Sounds good.

On replacing the separate clang path with a single toolchain path, we currently [juggle a couple different variables for compiling with clang to the target](https://github.com/apple/swift/blob/c6c88ce7140153ba828f580b821038c74a32cff7/stdlib/public/SwiftShims/CMakeLists.txt#L83), as opposed to the `--host-cc` variable for the clang that compiles for the host. We should make sure that keeps working.

I like the existing flexibility of having separate paths for the target clang toolchain and Swift toolchain, say if you're cross-compiling and have to use a separate C++ cross-compilation toolchain for your target, and I have used it that way in the past but I don't currently do so, as the Concurrency runtime has to built by the Swift-forked clang.

Reply:
> the runtimes contain swiftcc/swift-asynccc, so it is unsafe to use an unmatched compiler to build those.

You are right that it is no longer something that can be done, though [I used to cross-compile the C++ files in the stdlib for Android with the Android NDK clang before](https://github.com/buttaface/swift-android-sdk/commit/bd7e6746cff3f9cc25b836ba481dc11e7ee62840#diff-76992bc322b65ab94f299c05ba3cbd8b3ab2e32ebfdde6db5adba70fa9f5a26aL132).

No need to add those additional build flags for me, as Android is already special-cased anyway.